[WIP] Implement conditionals.

jmchilton opened the pull request:
```
sh run_tests.sh -api lib/galaxy_test/api/cwl/test_cwl_conformance_red_v1_2.py::CwlConformanceTestCase::test_conformance_v1_2_direct_optional_null_result
```

jmchilton committed:
Add CWL conformance test data to Galaxy for testing.

Co-authored-by: Nicola Soranzo <nicola.soranzo@earlham.ac.uk>

jmchilton committed:
Allow unit testing tool init to specify paths.

jmchilton committed:
More logging messages... REMOVE LATER

jmchilton committed:
WIP: Work toward Galaxy-flavored CWL tools.

gx:Interface.

Co-authored-by: John Chilton <jmchilton@gmail.com>
Co-authored-by: jra001

jmchilton committed:
CWL Testing and Runner Improvements.

jmchilton committed:
Revert complex objectstore - for docker tests (needed for CWL).

jmchilton committed:
[WIP] Implement client UI for field parameter type for CWL.

jmchilton committed:
Deal with workflow definitions without position fields.

jmchilton committed:
Swap default for beta formats (do not include in Galaxy).

jmchilton committed:
[WIP] Implement records - heterogenous dataset collections.

Existing dataset colleciton types are meant to be homogenous - all datasets of the same time. This introduces CWL-style record dataset collections.

jmchilton committed:
Implement subset of the Common Workflow Language tool and workflow formats.

This should support a subset of [draft-3](http://www.commonwl.org/draft-3/) and [v1.0](http://www.commonwl.org/v1.0/) tools.

CWL Support (Tools):
--------------------

- Implemented integer, long, float, double, boolean, string, File, Directory, "null", Any, as well as records and arrays thereof. There are two approaches to handling more complex parameters discussed here (https://github.com/common-workflow-language/galaxy/issues/59).
- ``secondaryFiles`` that are actual Files are implemented, secondaryFiles containing directories are not yet implemented.
- ``InlineJavascriptRequirement`` are support to define output files (see ``test_cat3`` test case).
- ``EnvVarRequirement``s are supported (see the ``test_env_tool1`` and ``test_env_tool2`` test cases).
- Expression tools are supported (see ``parseInt-tool`` test case).
- Shell tools are also support (see record output test case).
- Default File values are very un-Galaxy and have been hacked into work with Tools - they still don't work with workflows.
- Partial Docker support - this supports the most simple and common pullFrom semantics but not additional ways to fetch containers or additional options such as output directory configuration (https://github.com/common-workflow-language/galaxy/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20Docker). Additionally, Galaxy mounts the inputs and outputs where it wants instead of CWL required mount points - this needs to be fixed for the conformance tests but may not matter much in practice (I'm not sure).

CWL Support (Workflows):
------------------------

- Simple connections and tool execution.
- Overriding tool input defaults via literal values and simple expressions.
- MultipleInputFeatureRequirements to glue together multiple file inputs into a File[] or multiple File[] into a single flat File[]. (nested merge is still a TODO).
- Simple scatter semantics for Files and non-Files (e.g. count-lines3).
- Simple subworkflows (e.g. count-lines10).
- Simple valueFrom expressions (e.g. ``step-valueFrom`` and ``step-valueFrom2``). This work doesn't yet model non-tool parameters to steps - for complex ``valueFrom`` expressions like in ``step-valueFrom3`` do not work yet.

Remaining Work
---------------------------------

The work remaining is vast and will be tracked at https://github.com/common-workflow-language/galaxy/issues for the time being.

Implementation Notes:
----------------------

Tools:

 - Non-File CWL outputs are represented as ``expression.json`` files. Traditionally Galaxy hasn't supported non-File outputs from tools but CWL Galaxy has work in progress on bringing native Galaxy support for such outputs https://github.com/common-workflow-language/galaxy/issues/27.
 - CWL secondary files are just normal datasets with extra files stored in ``__secondary_files__`` directory in the dataset's extra_files_path directory and indexed in a file called __secondary_files_index.json in extra_files_path. The upload tools has been augmented to allow attaching arbitrary extra files as a tar file to support getting data into this format initially. CWL requires staging files to include their parent File's ``basename`` - but tools describe inputs as just the extension. I'm not sure which way Galaxy should store __secondary_files__ in its objectstore - just with the extension or with the basename and extension - both options are implemented and can be swapped by setting the boolean STORE_SECONDARY_FILES_WITH_BASENAME in galaxy.tools.cwl.util.
 - CWL Directory types are datasets of a new type "directory" implemented earlier in this branch.
 - The tool execution API has been extended to add a ``inputs_representation`` parameter that can be set to "cwl" now. The ``cwl`` representation for running tools corresonding to the CWL job json format with {class: "File: path: "/path/to/file"} inputs replaced with {"src": "hda", "id": "<dataset_id>"}. Code for building these requests for CWL job json is available in the test class.
 - Since the CWL <-> Galaxy parameter translation may change over time, for instance if Galaxy develops or refines parameter classes - CWL state and CWL state version is tracked in the database and hopefully for reruns, etc... we could update the Galaxy state from an older version to a new one.
 - CWL allows output parameters to be either ``File`` or non-``File`` and determined at runtime, so ``galaxy.json`` is used to dynamically adjust output extension as needed for non-``File`` parameters.

Workflows:

- This work serializes embedded and referenced tools into the database - this will allow reuse and tracing without require the path to exist forever on the filesystem - this will have problems with default file references in workflows.
- Implements re-mapping CWL workflow connections to Galaxy input connections.
- Fix tool serialization for jobs for path-less tools (such as embedded tools).
- Hack tool state during workflow import for CWL.
- The sort of dynamic shaping of inputs CWL allows has required enhancing Galaxy's map/reduce stuff to allow mapping over dynamic collections that don't yet exist at the time of tool execution and need to be created on the fly. This commit creates them as HDCAs - but likely they should be something else that doesn't appear in the history panel.
- Multi-input scattering but only scatterMethod == "dotproduct" is currently support. Other scatter methods (nested_crossproduct and flatcross_product) are not used by workflows in GA4GH challenge.

Implementation Description:
-----------------------------

The reference implementation Python library (mainly developed by Peter Amstutz - https://github.com/common-workflow-language/common-workflow-language/tree/master/reference) is used to load tool files ending with ``.json`` or ``.cwl`` and proxy objects are created to adapt these tools to Galaxy representations. In particular input and output descriptions are loaded from the tool.

When the tool is submitted, a special specialized tool class is used to build a cwltool compatible job description from the supplied Galaxy inputs and the CWL reference implementation is used to generate a CWL reference implementation Job object. A command-line is generated from this Job object.

As a result of this - Galaxy largely does not need to worry about the details of command-line adapters, expressions, etc....

Galaxy writes a description of the CWL job that it can reload to the job working directory. After the process is complete (on the Galaxy compute server, but outside the Docker container) this representation is reloaded and the dynamic outputs are discovered and moved to fixed locations as expected by Galaxy. CWL allows for much more expressive output locations than Galaxy, for better or worse, and this step uses cwltool to adapt CWL to Galaxy outputs.

Currently all ``File`` outputs are sniffed to determined a Galaxy datatype, CWL allows refinement on this and this remains work to be done.

  1) CWL should support EDAM declaration of types and Galaxy should provide a mapping to core datasets to skip sniffing is types are found.
  2) For finer grain control within Galaxy, extensions to CWL should allow setting actual Galaxy output types on outputs. (Distinction between fastq and fastqsanger in Galaxy is very important for instance.)

Implementation Links:
----------------------

Hundreds of commits have been rebased into this one and so the details of individual parts of the implementation and how they built on each other are not enitrely clear. To see the original ideas behind individual features - here are some relevant links:

- Implement merge_nested link semantics for workflow steps (https://github.com/common-workflow-language/galaxy/commit/a903abdb65e54d11d1f25d732f9e04da15cc1ff6).
- Implement subworkflows in CWL (https://github.com/common-workflow-language/galaxy/commit/9933c3c211b93b328cf63afba66f144842ff9e10)
- MultipleInputFeatureRequirements:
  - Second attempt: https://github.com/common-workflow-language/galaxy/commit/ed8307f5c3a5e8ea4318807c893543c45fa72cf7
  - First attempt: https://github.com/common-workflow-language/galaxy/commit/ae11f5662cd5254bde1c1cb345d780be8317f59b
- Basic, implicit dotproduct scattering of workflows - https://github.com/common-workflow-language/galaxy/commit/d1ad64ed6f2e000ae1d261c7c22f54ab3b9dcb4d.
- Simple input StepInputExpressionRequirements - https://github.com/common-workflow-language/galaxy/commit/819a27b816ecc71c2190efb33b9127658667e784
- StepInputExpressionRequirements for multiple inputs - https://github.com/common-workflow-language/galaxy/commit/5e7f622b4ff072eafa844af525f4367de5e7e644
- Record Types in CWL - https://github.com/common-workflow-language/galaxy/commit/e6be28a2d4983184757aafd1fafdb1b56b692438
- Rework original approach at mapping CWL state to tool state - https://github.com/common-workflow-language/galaxy/commit/669ea55a7ab55e8741a82d0dcd5a15f5075358dd
- Rework approach at mapping CWL state to tool state again to use "FieldTypeToolParameter"s - implements default values, optional parameters, and union types for workflow inputs. https://github.com/common-workflow-language/galaxy/commit/d1ca22fab5ca8adb8691b2f76846c41aaa798e49
- Initial tracking of "cwl_filename" for CWL jobs (https://github.com/common-workflow-language/galaxy/commit/67ffc5573f45d8293f9a4df502c56247964e11ab).
- Reworked secondary file staging, implement testing and indexing of secondary files - https://github.com/common-workflow-language/galaxy/commit/03d1636fc2eb7ba4e5b42526ea69cce447252c46.

Testing:
---------------------

    % git clone https://github.com/common-workflow-language/galaxy.git
    % cd galaxy
    % git checkout cwl-1.0

Start Galaxy.

    % GALAXY_RUN_WITH_TEST_TOOLS=1 sh run.sh

Open http://localhost:8080/ and see CWL test tools (along with all Galaxy test tools) in left hand tool panel.

To go a step further and actually run CWL jobs within their designated Docker containers, copy the following minimal Galaxy job configuration file to ``config/job_conf.xml``. (Adjust the ``docker_sudo`` parameter based on how you execute Docker).

https://gist.github.com/jmchilton/3997fa471d1b4c556966

Run API tests demonstrating the various CWL demo tools with the following command.

```
./run_tests.sh -api test/api/test_tools_cwl.py
./run_tests.sh -api test/api/test_workflows_cwl.py
./run_tests.sh -api test/api/test_cwl_conformance_v1_0.py
```

An individual conformance test can be ran using this pattern:

```
./run_tests.sh -api test/api/test_cwl_conformance_v1_0.py:CwlConformanceTestCase.test_conformance_v1_0_6
```

The first two execute various tool and workflow test cases manually crafted during implementation of this work. The third is an auto-generate test case class that contains Python tests for every CWL conformance test found with the reference specification.

Issues and Contact
---------------------------------

Report issues at https://github.com/common-workflow-language/galaxy/issues and feel free ping jmchilton on the CWL [Gitter channel](https://gitter.im/common-workflow-language/common-workflow-language).

Co-authored-by: Hervé MENAGER <herve.menager@pasteur.fr>
Co-authored-by: John Chilton <jmchilton@gmail.com>
Co-authored-by: Michael R. Crusoe <michael.crusoe@gmail.com>
Co-authored-by: jra001k
Co-authored-by: mvdbeek <m.vandenbeek@gmail.com>
Co-authored-by: Nicola Soranzo <nicola.soranzo@earlham.ac.uk>

jmchilton committed:
WORKAROUND TO GET TAR TO DIRECTORY WORKING AGAIN.

jmchilton commented:
03c56063be6f5c7b03d75ff53e1b3961dde2515a is the actual conditional implementing, and I think it passes every test case that includes a when with no ``pickValues``!

I was remembering pickValues being on inputs... but I was clearly wrong. All the test cases have them on workflow outputs and that is the only place they can be according to the spec. This makes implementing them more tricky for sure - some ideas for paths forward on https://github.com/common-workflow-language/galaxy/issues/126

mvdbeek commented on lib/galaxy/workflow/modules.py:
```diff
@@ -1760,6 +1761,23 @@ def from_cwl(value):
             step_state[key] = to_cwl(value)
 
+        if when_expression is not None:
+            from cwltool.expression import do_eval
```
Let's import this at the top

jmchilton replied:
You can tell I have developed habits stemming from frequent rebasing, imports at top result in conflicts 😆 .

mr-c commented:
Hey @jmchilton , what can we do to get this merged?

mr-c commented:
I guess this PR can get rebased or merged with https://github.com/galaxyproject/galaxy/pull/10712 now that it has been accepted upstream?

nsoranzo committed:
Generate Python test cases from CWL conformance tests

mr-c committed:
Create codeql-analysis.yml

jmchilton committed:
[WIP] Implement conditionals.

```
sh run_tests.sh -api lib/galaxy_test/api/cwl/test_cwl_conformance_red_v1_2.py::CwlConformanceTestCase::test_conformance_v1_2_direct_optional_null_result
```

mr-c commented:
Already merged!
